pkg/lease/client.go

```diff
@@ -0,0 +1,99 @@
+package lease
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	boskos "k8s.io/test-infra/boskos/client"
+	"k8s.io/test-infra/boskos/common"
+)
+
+const (
+	freeState   = "free"
+	leasedState = "leased"
+)
+
+type boskosClient interface {
+	Acquire(rtype, state, dest string) (*common.Resource, error)
+	UpdateOne(name, dest string, _ *common.UserData) error
+	ReleaseOne(name, dest string) error
+	ReleaseAll(dest string) error
+}
+
+type Client interface {
+	Acquire(rtype string, cancel context.CancelFunc) (string, error)
+	Heartbeat() error
+	Release(name string) error
+	ReleaseAll() ([]string, error)
+}
+
+func NewClient(owner, url string) Client {
+	return newClient(boskos.NewClient(owner, url))
+}
+
+func newClient(boskos boskosClient) Client {
+	return &client{
+		boskos: boskos,
+		cancel: make(map[string]context.CancelFunc),
+	}
+}
+
+type client struct {
+	sync.RWMutex
+	boskos boskosClient
+	cancel map[string]context.CancelFunc
+}
+
+func (c *client) Acquire(rtype string, cancel context.CancelFunc) (string, error) {
+	c.Lock()
```

Comment:
technically you only need to lock around `c.cancel` update, right?

Reply:
I missed that after the transition from `ReleaseAll` to `ReleaseOne`. Changed `Release` as well.

Also fixed `Heartbeat` to use a write lock since it now also mutates the client.